{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6814f4a0",
   "metadata": {},
   "source": [
    "# 2. Understanding multicellular programs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e923b97",
   "metadata": {},
   "source": [
    "### Loading data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c1493c5",
   "metadata": {},
   "source": [
    "ReCoN generally uses single-cell RNA-seq data and/or single-cell ATAC-seq data.\n",
    "```{tip}\n",
    "The data for this tutorial can be downloaded together at _____.\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fc9cbd89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/pasteur/helix/projects/ml4ig_hot/Users/rtrimbou/ReCoN\n"
     ]
    }
   ],
   "source": [
    "cd /pasteur/helix/projects/ml4ig_hot/Users/rtrimbou/ReCoN/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e23f1b17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mWARNING: Skipping recon as it is not installed.\u001b[0m\u001b[33m\n",
      "\u001b[0mNote: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip uninstall recon -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cf1d16cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing /pasteur/helix/projects/ml4ig_hot/Users/rtrimbou/ReCoN\n",
      "  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: numpy<2.0.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from recon==0.1.0) (1.26.4)\n",
      "Requirement already satisfied: pandas in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from recon==0.1.0) (2.3.3)\n",
      "Collecting hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy (from recon==0.1.0)\n",
      "  Cloning https://github.com/cantinilab/hummus (to revision dask_update) to /local/scratch/tmp/pip-install-qn35djbp/hummuspy_8028b1ad78414a96942a6c9aea56edb3\n",
      "  Running command git clone --filter=blob:none --quiet https://github.com/cantinilab/hummus /local/scratch/tmp/pip-install-qn35djbp/hummuspy_8028b1ad78414a96942a6c9aea56edb3\n",
      "  Running command git checkout -b dask_update --track origin/dask_update\n",
      "  Switched to a new branch 'dask_update'\n",
      "  branch 'dask_update' set up to track 'origin/dask_update'.\n",
      "  Resolved https://github.com/cantinilab/hummus to commit 8dff0c239882d752d9c131d208436d6b30bef717\n",
      "  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: multixrank in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from recon==0.1.0) (0.3)\n",
      "Requirement already satisfied: scanpy in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from recon==0.1.0) (1.11.5)\n",
      "Requirement already satisfied: scipy in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from recon==0.1.0) (1.15.3)\n",
      "Requirement already satisfied: plotly in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from recon==0.1.0) (6.5.1)\n",
      "Requirement already satisfied: arboreto in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from recon==0.1.0) (0.1.6)\n",
      "Requirement already satisfied: matplotlib in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from recon==0.1.0) (3.10.8)\n",
      "Requirement already satisfied: circe-py>=0.3.9 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from recon==0.1.0) (0.3.9)\n",
      "Requirement already satisfied: bokeh!=3.0.*,>=2.4.2 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (3.8.2)\n",
      "Requirement already satisfied: dask>=2023.0.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (2025.12.0)\n",
      "Requirement already satisfied: distributed>=2023.0.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (2025.12.0)\n",
      "Requirement already satisfied: joblib<2.0.0,>=1.3.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (1.5.3)\n",
      "Requirement already satisfied: pyyaml<7.0,>=6.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (6.0.3)\n",
      "Requirement already satisfied: rich>=10.12.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (14.2.0)\n",
      "Requirement already satisfied: tqdm>=4.66.3 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (4.67.1)\n",
      "Requirement already satisfied: celloracle-lite>=0.21.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from recon==0.1.0) (0.21.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib->recon==0.1.0) (1.3.2)\n",
      "Requirement already satisfied: cycler>=0.10 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib->recon==0.1.0) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib->recon==0.1.0) (4.61.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib->recon==0.1.0) (1.4.9)\n",
      "Requirement already satisfied: packaging>=20.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib->recon==0.1.0) (25.0)\n",
      "Requirement already satisfied: pillow>=8 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib->recon==0.1.0) (12.1.0)\n",
      "Requirement already satisfied: pyparsing>=3 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib->recon==0.1.0) (3.3.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib->recon==0.1.0) (2.9.0.post0)\n",
      "Requirement already satisfied: networkx in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from multixrank->recon==0.1.0) (3.4.2)\n",
      "Requirement already satisfied: psutil in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from multixrank->recon==0.1.0) (7.2.1)\n",
      "Requirement already satisfied: pytz>=2020.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from pandas->recon==0.1.0) (2025.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from pandas->recon==0.1.0) (2025.3)\n",
      "Requirement already satisfied: Jinja2>=2.9 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from bokeh!=3.0.*,>=2.4.2->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (3.1.6)\n",
      "Requirement already satisfied: narwhals>=1.13 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from bokeh!=3.0.*,>=2.4.2->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (2.15.0)\n",
      "Requirement already satisfied: tornado>=6.2 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from bokeh!=3.0.*,>=2.4.2->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (6.5.4)\n",
      "Requirement already satisfied: xyzservices>=2021.09.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from bokeh!=3.0.*,>=2.4.2->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (2025.11.0)\n",
      "Requirement already satisfied: cython in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from celloracle-lite>=0.21.0->recon==0.1.0) (3.2.4)\n",
      "Requirement already satisfied: numba>=0.50.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from celloracle-lite>=0.21.0->recon==0.1.0) (0.63.1)\n",
      "Requirement already satisfied: scikit-learn in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from celloracle-lite>=0.21.0->recon==0.1.0) (1.7.2)\n",
      "Requirement already satisfied: IPython in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from celloracle-lite>=0.21.0->recon==0.1.0) (8.37.0)\n",
      "Requirement already satisfied: anndata<=0.10.8,>=0.7.5 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from celloracle-lite>=0.21.0->recon==0.1.0) (0.10.8)\n",
      "Requirement already satisfied: genomepy>=0.8.4 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from celloracle-lite>=0.21.0->recon==0.1.0) (0.16.3)\n",
      "Requirement already satisfied: gimmemotifs<=0.17.2,>=0.14.4 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from celloracle-lite>=0.21.0->recon==0.1.0) (0.17.2)\n",
      "Requirement already satisfied: statsmodels==0.14.5 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from celloracle-lite>=0.21.0->recon==0.1.0) (0.14.5)\n",
      "Requirement already satisfied: patsy>=0.5.6 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from statsmodels==0.14.5->celloracle-lite>=0.21.0->recon==0.1.0) (1.0.2)\n",
      "Requirement already satisfied: array-api-compat!=1.5,>1.4 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from anndata<=0.10.8,>=0.7.5->celloracle-lite>=0.21.0->recon==0.1.0) (1.13.0)\n",
      "Requirement already satisfied: exceptiongroup in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from anndata<=0.10.8,>=0.7.5->celloracle-lite>=0.21.0->recon==0.1.0) (1.3.1)\n",
      "Requirement already satisfied: h5py>=3.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from anndata<=0.10.8,>=0.7.5->celloracle-lite>=0.21.0->recon==0.1.0) (3.15.1)\n",
      "Requirement already satisfied: natsort in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from anndata<=0.10.8,>=0.7.5->celloracle-lite>=0.21.0->recon==0.1.0) (8.4.0)\n",
      "Requirement already satisfied: setuptools>=0.7 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (80.9.0)\n",
      "Requirement already satisfied: biofluff>=3.0.4 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (3.0.4)\n",
      "Requirement already satisfied: configparser in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (7.2.0)\n",
      "Requirement already satisfied: diskcache in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (5.6.3)\n",
      "Requirement already satisfied: feather-format in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (0.4.1)\n",
      "Requirement already satisfied: iteround in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (1.0.4)\n",
      "Requirement already satisfied: logomaker in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (0.8.7)\n",
      "Requirement already satisfied: loguru in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (0.7.3)\n",
      "Requirement already satisfied: pybedtools>=0.9.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (0.12.0)\n",
      "Requirement already satisfied: pysam>=0.16 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (0.23.3)\n",
      "Requirement already satisfied: xxhash in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (3.6.0)\n",
      "Requirement already satisfied: qnorm>=0.8.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (0.9.0)\n",
      "Requirement already satisfied: seaborn>=0.10.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (0.13.2)\n",
      "Requirement already satisfied: xdg in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (6.0.0)\n",
      "Requirement already satisfied: HTSeq in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from biofluff>=3.0.4->gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (2.0.9)\n",
      "Requirement already satisfied: palettable in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from biofluff>=3.0.4->gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (3.3.3)\n",
      "Requirement already satisfied: pyBigWig in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from biofluff>=3.0.4->gimmemotifs<=0.17.2,>=0.14.4->celloracle-lite>=0.21.0->recon==0.1.0) (0.3.24)\n",
      "Requirement already satisfied: attrs>=20.3 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from circe-py>=0.3.9->recon==0.1.0) (25.4.0)\n",
      "Requirement already satisfied: click>=8.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from dask>=2023.0.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (8.3.1)\n",
      "Requirement already satisfied: cloudpickle>=3.0.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from dask>=2023.0.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (3.1.2)\n",
      "Requirement already satisfied: fsspec>=2021.09.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from dask>=2023.0.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (2026.1.0)\n",
      "Requirement already satisfied: partd>=1.4.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from dask>=2023.0.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (1.4.2)\n",
      "Requirement already satisfied: toolz>=0.12.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from dask>=2023.0.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (1.1.0)\n",
      "Requirement already satisfied: importlib_metadata>=4.13.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from dask>=2023.0.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (8.7.1)\n",
      "Requirement already satisfied: locket>=1.0.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from distributed>=2023.0.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (1.0.0)\n",
      "Requirement already satisfied: msgpack>=1.0.2 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from distributed>=2023.0.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (1.1.2)\n",
      "Requirement already satisfied: sortedcontainers>=2.0.5 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from distributed>=2023.0.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (2.4.0)\n",
      "Requirement already satisfied: tblib!=3.2.0,!=3.2.1,>=1.6.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from distributed>=2023.0.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (3.2.2)\n",
      "Requirement already satisfied: urllib3>=1.26.5 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from distributed>=2023.0.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (2.6.3)\n",
      "Requirement already satisfied: zict>=3.0.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from distributed>=2023.0.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (3.0.0)\n",
      "Requirement already satisfied: appdirs in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (1.4.4)\n",
      "Requirement already satisfied: biopython>=1.73 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (1.86)\n",
      "Requirement already satisfied: colorama in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (0.4.6)\n",
      "Requirement already satisfied: filelock>=3.5 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (3.20.3)\n",
      "Requirement already satisfied: mygene in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (3.2.2)\n",
      "Requirement already satisfied: mysql-connector-python<=8.4 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (8.4.0)\n",
      "Requirement already satisfied: pyfaidx>=0.7.2.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (0.9.0.3)\n",
      "Requirement already satisfied: requests in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (2.32.5)\n",
      "Requirement already satisfied: zipp>=3.20 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from importlib_metadata>=4.13.0->dask>=2023.0.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (3.23.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from Jinja2>=2.9->bokeh!=3.0.*,>=2.4.2->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (3.0.3)\n",
      "Requirement already satisfied: llvmlite<0.47,>=0.46.0dev0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from numba>=0.50.1->celloracle-lite>=0.21.0->recon==0.1.0) (0.46.0)\n",
      "Requirement already satisfied: six>=1.5 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from python-dateutil>=2.7->matplotlib->recon==0.1.0) (1.17.0)\n",
      "Requirement already satisfied: markdown-it-py>=2.2.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from rich>=10.12.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (4.0.0)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from rich>=10.12.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (2.19.2)\n",
      "Requirement already satisfied: mdurl~=0.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from markdown-it-py>=2.2.0->rich>=10.12.0->hummuspy@ git+https://github.com/cantinilab/hummus@dask_update#subdirectory=hummuspy->recon==0.1.0) (0.1.2)\n",
      "Requirement already satisfied: legacy-api-wrap>=1.4.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from scanpy->recon==0.1.0) (1.5)\n",
      "Requirement already satisfied: pynndescent>=0.5.13 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from scanpy->recon==0.1.0) (0.6.0)\n",
      "Requirement already satisfied: session-info2 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from scanpy->recon==0.1.0) (0.3)\n",
      "Requirement already satisfied: typing-extensions in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from scanpy->recon==0.1.0) (4.15.0)\n",
      "Requirement already satisfied: umap-learn>=0.5.6 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from scanpy->recon==0.1.0) (0.5.9.post2)\n",
      "Requirement already satisfied: threadpoolctl>=3.1.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from scikit-learn->celloracle-lite>=0.21.0->recon==0.1.0) (3.6.0)\n",
      "Requirement already satisfied: pyarrow>=14.0.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from dask[complete]->arboreto->recon==0.1.0) (22.0.0)\n",
      "Requirement already satisfied: lz4>=4.3.2 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from dask[complete]->arboreto->recon==0.1.0) (4.4.5)\n",
      "Requirement already satisfied: decorator in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from IPython->celloracle-lite>=0.21.0->recon==0.1.0) (5.2.1)\n",
      "Requirement already satisfied: jedi>=0.16 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from IPython->celloracle-lite>=0.21.0->recon==0.1.0) (0.19.2)\n",
      "Requirement already satisfied: matplotlib-inline in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from IPython->celloracle-lite>=0.21.0->recon==0.1.0) (0.2.1)\n",
      "Requirement already satisfied: pexpect>4.3 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from IPython->celloracle-lite>=0.21.0->recon==0.1.0) (4.9.0)\n",
      "Requirement already satisfied: prompt_toolkit<3.1.0,>=3.0.41 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from IPython->celloracle-lite>=0.21.0->recon==0.1.0) (3.0.52)\n",
      "Requirement already satisfied: stack_data in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from IPython->celloracle-lite>=0.21.0->recon==0.1.0) (0.6.3)\n",
      "Requirement already satisfied: traitlets>=5.13.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from IPython->celloracle-lite>=0.21.0->recon==0.1.0) (5.14.3)\n",
      "Requirement already satisfied: wcwidth in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from prompt_toolkit<3.1.0,>=3.0.41->IPython->celloracle-lite>=0.21.0->recon==0.1.0) (0.2.14)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.4 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from jedi>=0.16->IPython->celloracle-lite>=0.21.0->recon==0.1.0) (0.8.5)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from pexpect>4.3->IPython->celloracle-lite>=0.21.0->recon==0.1.0) (0.7.0)\n",
      "Requirement already satisfied: biothings-client>=0.2.6 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from mygene->genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (0.4.1)\n",
      "Requirement already satisfied: httpx>=0.22.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from biothings-client>=0.2.6->mygene->genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (0.28.1)\n",
      "Requirement already satisfied: anyio in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from httpx>=0.22.0->biothings-client>=0.2.6->mygene->genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (4.12.1)\n",
      "Requirement already satisfied: certifi in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from httpx>=0.22.0->biothings-client>=0.2.6->mygene->genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (2026.1.4)\n",
      "Requirement already satisfied: httpcore==1.* in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from httpx>=0.22.0->biothings-client>=0.2.6->mygene->genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (1.0.9)\n",
      "Requirement already satisfied: idna in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from httpx>=0.22.0->biothings-client>=0.2.6->mygene->genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (3.11)\n",
      "Requirement already satisfied: h11>=0.16 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from httpcore==1.*->httpx>=0.22.0->biothings-client>=0.2.6->mygene->genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (0.16.0)\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from requests->genomepy>=0.8.4->celloracle-lite>=0.21.0->recon==0.1.0) (3.4.4)\n",
      "Requirement already satisfied: executing>=1.2.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from stack_data->IPython->celloracle-lite>=0.21.0->recon==0.1.0) (2.2.1)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from stack_data->IPython->celloracle-lite>=0.21.0->recon==0.1.0) (3.0.1)\n",
      "Requirement already satisfied: pure_eval in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from stack_data->IPython->celloracle-lite>=0.21.0->recon==0.1.0) (0.2.3)\n",
      "Building wheels for collected packages: recon\n",
      "  Building wheel for recon (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for recon: filename=recon-0.1.0-py3-none-any.whl size=394827812 sha256=dfa3fae3ca2fec554c18dbca20202d904d006b1778e50fdae145172d992ba0cb\n",
      "  Stored in directory: /local/scratch/tmp/pip-ephem-wheel-cache-1i0llchw/wheels/db/46/6c/6132d968d426d7198bf5fad84a78a2a9a7a6467a5434e3df7b\n",
      "Successfully built recon\n",
      "Installing collected packages: recon\n",
      "Successfully installed recon-0.1.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install .[grn-lite]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "702868f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting gseapy\n",
      "  Downloading gseapy-1.1.11-cp310-cp310-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (11 kB)\n",
      "Requirement already satisfied: numpy>=1.13.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gseapy) (1.26.4)\n",
      "Requirement already satisfied: scipy in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gseapy) (1.15.3)\n",
      "Requirement already satisfied: pandas in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gseapy) (2.3.3)\n",
      "Requirement already satisfied: matplotlib>=2.2 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gseapy) (3.10.8)\n",
      "Requirement already satisfied: requests in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from gseapy) (2.32.5)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib>=2.2->gseapy) (1.3.2)\n",
      "Requirement already satisfied: cycler>=0.10 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib>=2.2->gseapy) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib>=2.2->gseapy) (4.61.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib>=2.2->gseapy) (1.4.9)\n",
      "Requirement already satisfied: packaging>=20.0 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib>=2.2->gseapy) (25.0)\n",
      "Requirement already satisfied: pillow>=8 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib>=2.2->gseapy) (12.1.0)\n",
      "Requirement already satisfied: pyparsing>=3 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib>=2.2->gseapy) (3.3.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from matplotlib>=2.2->gseapy) (2.9.0.post0)\n",
      "Requirement already satisfied: six>=1.5 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from python-dateutil>=2.7->matplotlib>=2.2->gseapy) (1.17.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from pandas->gseapy) (2025.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from pandas->gseapy) (2025.3)\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from requests->gseapy) (3.4.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from requests->gseapy) (3.11)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from requests->gseapy) (2.6.3)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages (from requests->gseapy) (2026.1.4)\n",
      "Downloading gseapy-1.1.11-cp310-cp310-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (605 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m605.7/605.7 kB\u001b[0m \u001b[31m36.9 MB/s\u001b[0m  \u001b[33m0:00:00\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: gseapy\n",
      "Successfully installed gseapy-1.1.11\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install gseapy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a4c8d838",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scanpy as sc  # single cell data\n",
    "import pandas as pd  # data manipulation\n",
    "import liana as li  # cell communication\n",
    "import recon  # multilayer and perturbation prediction\n",
    "import recon.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4a010708",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna = sc.read_h5ad(\"./data/perturbation_tuto/rna.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e36755c",
   "metadata": {},
   "source": [
    "Let's check what cell types are present in this dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1b06d52b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['B_cell', 'ILC', 'Macrophage', 'MigDC', 'Monocyte']"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna.obs[\"celltype\"].unique().tolist()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bb18a92",
   "metadata": {},
   "source": [
    "## Create ReCoN's multilayer network"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "554f608c",
   "metadata": {},
   "source": [
    "### Importing GRNs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dfa4b0c",
   "metadata": {},
   "source": [
    "You can either generate GRNs direclty with ReCoN or import a previously generated one.<br>\n",
    "\n",
    "```{tip}\n",
    "If you wish to generate it directly with ReCoN, please follow the turorial [________]_______.\n",
    "```\n",
    "\n",
    "```{warning}\n",
    "You'll then require a python=3.10 conda environment, cf [Installation]_____.\n",
    "```\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "535370ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pax5</td>\n",
       "      <td>Mbd1</td>\n",
       "      <td>0.000095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Pax5</td>\n",
       "      <td>Smad1</td>\n",
       "      <td>0.000092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pax5</td>\n",
       "      <td>Smad5</td>\n",
       "      <td>0.000092</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source target    weight\n",
       "0   Pax5   Mbd1  0.000095\n",
       "1   Pax5  Smad1  0.000092\n",
       "2   Pax5  Smad5  0.000092"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grn_path = \"./data/perturbation_tuto/grn.csv\"\n",
    "grn = pd.read_csv(grn_path, index_col=0)\n",
    "grn = grn.sort_values(by=\"weight\", ascending=False)[:500_000]\n",
    "grn[\"source\"] = grn[\"source\"].str.capitalize()\n",
    "grn[\"target\"] = grn[\"target\"].str.capitalize()\n",
    "grn.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48defe91",
   "metadata": {},
   "source": [
    "### Computing cell communication"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc22840c",
   "metadata": {},
   "source": [
    "The cell-cell communication is inferred through LIANA+, an external pakage dedicatedto this task\n",
    "\n",
    "```{tip}\n",
    "For information, you can check LIANA+ documentation: https://liana-py.readthedocs.io/en/latest/\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7a1a521d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using resource `mouseconsensus`.\n",
      "Using `.X`!\n",
      "/pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages/anndata/_core/anndata.py:430: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "15364 features of mat are empty, they will be removed.\n",
      "Make sure that normalized counts are passed!\n",
      "/pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages/liana/method/_pipe_utils/_pre.py:146: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "/pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages/liana/method/_pipe_utils/_pre.py:149: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "0.36 of entities in the resource are missing from the data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating ligand-receptor stats for 1296 samples and 937 features\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [00:01<00:00, 627.52it/s]\n"
     ]
    }
   ],
   "source": [
    "li.method.cellphonedb(rna, \n",
    "            # NOTE by default the resource uses HUMAN gene symbols\n",
    "            resource_name=\"mouseconsensus\",\n",
    "            expr_prop=0.00,\n",
    "            use_raw=False,\n",
    "            groupby=\"celltype\",\n",
    "            verbose=True, key_added='cpdb_res')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93b07016",
   "metadata": {},
   "source": [
    "```{warning}\n",
    "ReCoN simply requires to rename the columns of the output dataframe of LIANA.\n",
    "```\n",
    "\n",
    "We rename ligand and receptors as 'source' and 'ligands', connected cell types as 'celltype_source' and 'celltype_target', and the scores as 'weight'.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7368cd92",
   "metadata": {},
   "outputs": [],
   "source": [
    "ccc_network = rna.uns[\"cpdb_res\"].copy()\n",
    "ccc_network = ccc_network[[\"ligand\", \"receptor\", \"lr_means\", \"source\", \"target\"]]\n",
    "ccc_network = ccc_network.rename(columns={\n",
    "    \"lr_means\": \"weight\",\n",
    "    \"source\": \"celltype_source\",\n",
    "    \"target\": \"celltype_target\",\n",
    "    \"ligand\": \"source\",\n",
    "    \"receptor\": \"target\"\n",
    "})\n",
    "ccc_network = ccc_network[ccc_network['weight'] != 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6216a45f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>weight</th>\n",
       "      <th>celltype_source</th>\n",
       "      <th>celltype_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>406685</th>\n",
       "      <td>App</td>\n",
       "      <td>Cd74</td>\n",
       "      <td>102.485008</td>\n",
       "      <td>cDC2</td>\n",
       "      <td>cDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405645</th>\n",
       "      <td>Copa</td>\n",
       "      <td>Cd74</td>\n",
       "      <td>102.370003</td>\n",
       "      <td>cDC1</td>\n",
       "      <td>cDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>410237</th>\n",
       "      <td>Copa</td>\n",
       "      <td>Cd74</td>\n",
       "      <td>102.366211</td>\n",
       "      <td>eTAC</td>\n",
       "      <td>cDC1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       source target      weight celltype_source celltype_target\n",
       "406685    App   Cd74  102.485008            cDC2            cDC1\n",
       "405645   Copa   Cd74  102.370003            cDC1            cDC1\n",
       "410237   Copa   Cd74  102.366211            eTAC            cDC1"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "ccc_network.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36d22d33",
   "metadata": {},
   "source": [
    "### Add receptor & receptor - target genes informations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "19ceb31a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1bg</td>\n",
       "      <td>Abca1</td>\n",
       "      <td>0.005156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1bg</td>\n",
       "      <td>Abcb1a</td>\n",
       "      <td>0.005877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1bg</td>\n",
       "      <td>Abcb1b</td>\n",
       "      <td>0.005877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A1bg</td>\n",
       "      <td>Acsl1</td>\n",
       "      <td>0.005915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A1bg</td>\n",
       "      <td>Adk</td>\n",
       "      <td>0.005092</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source  target    weight\n",
       "2   A1bg   Abca1  0.005156\n",
       "3   A1bg  Abcb1a  0.005877\n",
       "4   A1bg  Abcb1b  0.005877\n",
       "7   A1bg   Acsl1  0.005915\n",
       "8   A1bg     Adk  0.005092"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "receptor_genes = recon.data.load_data.load_receptor_genes(\"mouse_receptor_gene_from_NichenetPKN\")\n",
    "\n",
    "genes = np.unique(grn['source'].tolist() + grn['target'].tolist())\n",
    "receptor_genes = receptor_genes[receptor_genes['target'].isin(genes)]\n",
    "receptor_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d58e256",
   "metadata": {},
   "source": [
    "## Case 1: Multicellular program around a cell-type-specific pathway\n",
    "\n",
    "If you have a pathway of interest in a given cell type, you can predict the reaction of the other cell types t this activation.\n",
    "You can also look \"upstream\", to predict what and how the environemnt triggered the activation of your pathway of interest.\n",
    " \n",
    "```{tip}\n",
    "You can check the cardiac fibrosis showcase f our manuscript for an exende example :) []____\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef9a15d3",
   "metadata": {},
   "source": [
    "Lets start by choosing a gene set (e.g. a hallmark about TNF alpha activation), and a cell type hypothetically expressing it (e.g. macrophages)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "85e9a2c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['HALLMARK_ADIPOGENESIS', 'HALLMARK_ALLOGRAFT_REJECTION', 'HALLMARK_ANDROGEN_RESPONSE', 'HALLMARK_ANGIOGENESIS', 'HALLMARK_APICAL_JUNCTION']\n"
     ]
    }
   ],
   "source": [
    "import gseapy as gp\n",
    "from gseapy import Msigdb \n",
    "\n",
    "# we can use gseapy to download the hallmarks from MSigDB\n",
    "msig = Msigdb()\n",
    "hallmarks = msig.get_gmt(category='mh.all', dbver=\"2024.1.Mm\")\n",
    "print(list(hallmarks.keys())[:5])\n",
    "\n",
    "# Let's pick a hallmark gene set as seeds for our perturbation analysis,\n",
    "# and filter it to keep only genes present in our network\n",
    "gene_seeds = [gene for gene in hallmarks['HALLMARK_TNFA_SIGNALING_VIA_NFKB'] if gene in genes]\n",
    "\n",
    "# Create a dictionary with seed genes as keys and initial score 1, since all are equally important here\n",
    "gene_seeds = {seed:1 for seed in gene_seeds}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ab80f3e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "157"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gene_seeds)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7adb446e",
   "metadata": {},
   "source": [
    "We now, add the cell type information to these genes. Since here, all should be expressed in macrophage, we simply add '-macrophage' to all of them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "50d9f62e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Abca1::Macrophage': 1, 'Atf3::Macrophage': 1, 'Atp2b1::Macrophage': 1, 'B4galt1::Macrophage': 1, 'B4galt5::Macrophage': 1}\n"
     ]
    }
   ],
   "source": [
    "seeds = {f\"{gene}::Macrophage\": score for gene, score in gene_seeds.items() if score > 0}\n",
    "\n",
    "print({k: seeds[k] for k in list(seeds.keys())[:5]})  # print first 5 items"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3aab47c",
   "metadata": {},
   "source": [
    "```{warning}\n",
    "In ReCoN, molecules are named differently depending on the layer they belong to.\n",
    "In intracellular layers, molecules (e.g. TFs, target genes, receptors) are named by their gene symbol and cell type, separated by 2 double points (e.g. \"Nfkb1::Macrophage\").\n",
    "In extracellular layers, molecules (e.g. ligands, receptors) are named by their gene symbol and cell type, separated by a hyphen (e.g. \"CD40-Macrophage\").\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77006963",
   "metadata": {},
   "source": [
    "```{note}\n",
    "You can of course use your own gene sets as seeds, for example from your own differential expression\n",
    "analysis. If you do so, make sure to filter the genes to keep only those present in your network.\n",
    "Additionally, you can assign different scores to different genes (e.g. based on logFC or confidence).\n",
    "⚠️ All gene weights must be positive !\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2791a5d6",
   "metadata": {},
   "source": [
    "## Assemble the multicellular network"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5e46c4a",
   "metadata": {},
   "source": [
    "There is many modifiable parameters:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "ae814397",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_communication_graph_directed = False\n",
    "cell_communication_graph_weighted = True\n",
    "restart_proba = 0.6\n",
    "ccc_proba = 0.5\n",
    "grn_graph_weighted = True\n",
    "grn_graph_directed = False"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d432a449",
   "metadata": {},
   "source": [
    "receptor_layer = recon.layers.ReceptorLayer.from_receptor_genes(\n",
    "    receptor_genes,\n",
    "    directed=grn_graph_directed,\n",
    "    weighted=grn_graph_weighted\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "0a0cb39e",
   "metadata": {},
   "outputs": [],
   "source": [
    "receptor_layer = pd.DataFrame(\n",
    "    {\"source\": receptor_genes['source'].unique()+'_receptor',\n",
    "    \"target\": [\"fake_receptor\" for r in range(len(receptor_genes['source'].unique()))]\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "194026d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import recon.explore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "bb80d94b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages/recon/explore/recon.py:122: UserWarning: \n",
      "                No receptor_graph provided,\n",
      "                an empty receptor graph will be created.\n",
      "                \n",
      "/pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages/recon/explore/recon.py:370: UserWarning: The celltypes dictionary was converted toa list of Celltype objects.\n",
      "The keys of the dictionary will be the celltype names.\n"
     ]
    }
   ],
   "source": [
    "celltypes=[\"B_cell\", \"pDC\", \"Macrophage\", \"NK_cell\", \"T_cell_CD4\", \"T_cell_CD8\"]    # list of cell types to include in the analysis\n",
    "\n",
    "generic_multicell = recon.explore.Multicell(\n",
    "    celltypes = {celltype: recon.explore.Celltype(\n",
    "#        receptor_graph = receptor_layer,\n",
    "        grn_graph = grn,\n",
    "        receptor_grn_bipartite = receptor_genes,\n",
    "        celltype_name = celltype,\n",
    "        receptor_graph_directed=False,\n",
    "        receptor_graph_weighted=False,\n",
    "        grn_graph_directed=grn_graph_directed,\n",
    "        grn_graph_weighted=grn_graph_weighted,\n",
    "        receptor_grn_bipartite_graph_directed=False,\n",
    "        receptor_grn_bipartite_graph_weighted=True,\n",
    "        seeds = seeds)  # we can either pass a dictionary of Celltype objects, or build them on the fly\n",
    "        for celltype in celltypes},\n",
    "    cell_communication_graph = ccc_network.iloc[ccc_network[\"celltype_source\"].isin(celltypes).values & ccc_network[\"celltype_target\"].isin(celltypes).values, :],\n",
    "    cell_communication_graph_directed=cell_communication_graph_directed,\n",
    "    cell_communication_graph_weighted=cell_communication_graph_weighted,\n",
    "    # bipartite parameters can be -1, 0, 1 here\n",
    "    bipartite_grn_cell_communication_directed=False,\n",
    "    bipartite_grn_cell_communication_weighted=False,\n",
    "    bipartite_cell_communication_receptor_directed=False,\n",
    "    bipartite_cell_communication_receptor_weighted=False,\n",
    "    seeds = seeds,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccfa888e",
   "metadata": {},
   "source": [
    "Now, we need to precise the kind of exploration we want to do : upstream/downstream of gene set activation, and intracellular only or extracellular."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "d84654ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "generic_multicell.lamb = recon.explore.set_lambda(\n",
    "    generic_multicell,\n",
    "    direction=\"upstream\",\n",
    "    strategy=\"intercell\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51f8cac8",
   "metadata": {},
   "source": [
    "```{tip}\n",
    "Alternatively, you can modify lambda transition probabilities freely, to modulate GRN and CCC exploration\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "6d1864d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_communication</th>\n",
       "      <th>B_cell_receptor</th>\n",
       "      <th>B_cell_grn</th>\n",
       "      <th>pDC_receptor</th>\n",
       "      <th>pDC_grn</th>\n",
       "      <th>Macrophage_receptor</th>\n",
       "      <th>Macrophage_grn</th>\n",
       "      <th>NK_cell_receptor</th>\n",
       "      <th>NK_cell_grn</th>\n",
       "      <th>T_cell_CD4_receptor</th>\n",
       "      <th>T_cell_CD4_grn</th>\n",
       "      <th>T_cell_CD8_receptor</th>\n",
       "      <th>T_cell_CD8_grn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cell_communication</th>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.142857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B_cell_receptor</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B_cell_grn</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pDC_receptor</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pDC_grn</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Macrophage_receptor</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Macrophage_grn</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK_cell_receptor</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK_cell_grn</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T_cell_CD4_receptor</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T_cell_CD4_grn</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T_cell_CD8_receptor</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T_cell_CD8_grn</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     cell_communication  B_cell_receptor  B_cell_grn  \\\n",
       "cell_communication             0.142857              0.0    0.142857   \n",
       "B_cell_receptor                0.500000              0.5    0.000000   \n",
       "B_cell_grn                     0.000000              0.5    0.500000   \n",
       "pDC_receptor                   0.500000              0.0    0.000000   \n",
       "pDC_grn                        0.000000              0.0    0.000000   \n",
       "Macrophage_receptor            0.500000              0.0    0.000000   \n",
       "Macrophage_grn                 0.000000              0.0    0.000000   \n",
       "NK_cell_receptor               0.500000              0.0    0.000000   \n",
       "NK_cell_grn                    0.000000              0.0    0.000000   \n",
       "T_cell_CD4_receptor            0.500000              0.0    0.000000   \n",
       "T_cell_CD4_grn                 0.000000              0.0    0.000000   \n",
       "T_cell_CD8_receptor            0.500000              0.0    0.000000   \n",
       "T_cell_CD8_grn                 0.000000              0.0    0.000000   \n",
       "\n",
       "                     pDC_receptor   pDC_grn  Macrophage_receptor  \\\n",
       "cell_communication            0.0  0.142857                  0.0   \n",
       "B_cell_receptor               0.0  0.000000                  0.0   \n",
       "B_cell_grn                    0.0  0.000000                  0.0   \n",
       "pDC_receptor                  0.5  0.000000                  0.0   \n",
       "pDC_grn                       0.5  0.500000                  0.0   \n",
       "Macrophage_receptor           0.0  0.000000                  0.5   \n",
       "Macrophage_grn                0.0  0.000000                  0.5   \n",
       "NK_cell_receptor              0.0  0.000000                  0.0   \n",
       "NK_cell_grn                   0.0  0.000000                  0.0   \n",
       "T_cell_CD4_receptor           0.0  0.000000                  0.0   \n",
       "T_cell_CD4_grn                0.0  0.000000                  0.0   \n",
       "T_cell_CD8_receptor           0.0  0.000000                  0.0   \n",
       "T_cell_CD8_grn                0.0  0.000000                  0.0   \n",
       "\n",
       "                     Macrophage_grn  NK_cell_receptor  NK_cell_grn  \\\n",
       "cell_communication         0.142857               0.0     0.142857   \n",
       "B_cell_receptor            0.000000               0.0     0.000000   \n",
       "B_cell_grn                 0.000000               0.0     0.000000   \n",
       "pDC_receptor               0.000000               0.0     0.000000   \n",
       "pDC_grn                    0.000000               0.0     0.000000   \n",
       "Macrophage_receptor        0.000000               0.0     0.000000   \n",
       "Macrophage_grn             0.500000               0.0     0.000000   \n",
       "NK_cell_receptor           0.000000               0.5     0.000000   \n",
       "NK_cell_grn                0.000000               0.5     0.500000   \n",
       "T_cell_CD4_receptor        0.000000               0.0     0.000000   \n",
       "T_cell_CD4_grn             0.000000               0.0     0.000000   \n",
       "T_cell_CD8_receptor        0.000000               0.0     0.000000   \n",
       "T_cell_CD8_grn             0.000000               0.0     0.000000   \n",
       "\n",
       "                     T_cell_CD4_receptor  T_cell_CD4_grn  T_cell_CD8_receptor  \\\n",
       "cell_communication                   0.0        0.142857                  0.0   \n",
       "B_cell_receptor                      0.0        0.000000                  0.0   \n",
       "B_cell_grn                           0.0        0.000000                  0.0   \n",
       "pDC_receptor                         0.0        0.000000                  0.0   \n",
       "pDC_grn                              0.0        0.000000                  0.0   \n",
       "Macrophage_receptor                  0.0        0.000000                  0.0   \n",
       "Macrophage_grn                       0.0        0.000000                  0.0   \n",
       "NK_cell_receptor                     0.0        0.000000                  0.0   \n",
       "NK_cell_grn                          0.0        0.000000                  0.0   \n",
       "T_cell_CD4_receptor                  0.5        0.000000                  0.0   \n",
       "T_cell_CD4_grn                       0.5        0.500000                  0.0   \n",
       "T_cell_CD8_receptor                  0.0        0.000000                  0.5   \n",
       "T_cell_CD8_grn                       0.0        0.000000                  0.5   \n",
       "\n",
       "                     T_cell_CD8_grn  \n",
       "cell_communication         0.142857  \n",
       "B_cell_receptor            0.000000  \n",
       "B_cell_grn                 0.000000  \n",
       "pDC_receptor               0.000000  \n",
       "pDC_grn                    0.000000  \n",
       "Macrophage_receptor        0.000000  \n",
       "Macrophage_grn             0.000000  \n",
       "NK_cell_receptor           0.000000  \n",
       "NK_cell_grn                0.000000  \n",
       "T_cell_CD4_receptor        0.000000  \n",
       "T_cell_CD4_grn             0.000000  \n",
       "T_cell_CD8_receptor        0.000000  \n",
       "T_cell_CD8_grn             0.500000  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# layers = generic_multicell.lamb.index\n",
    "# is_grn = layers.str.endswith(\"_grn\")\n",
    "# generic_multicell.lamb.loc[is_grn, is_grn] = generic_multicell.lamb.loc[is_grn, is_grn]\n",
    "# generic_multicell.lamb.loc[is_grn, \"cell_communication\"] = generic_multicell.lamb.loc[is_grn, \"cell_communication\"]*1\n",
    "generic_multicell.lamb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b299e3e4",
   "metadata": {},
   "source": [
    "### We now run the analysis through random walks with restart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "8c9f0bb3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seeds are provided as a dictionary with weights per seed.\n",
      "Creating a multixrank object with seeds as a dictionary.\n",
      "cell_communication\n",
      "receptor\n",
      "gene\n",
      "receptor\n",
      "gene\n",
      "receptor\n",
      "gene\n",
      "receptor\n",
      "gene\n",
      "receptor\n",
      "gene\n",
      "receptor\n",
      "gene\n",
      "Identifying produced ligands in response to the perturbation.\n"
     ]
    }
   ],
   "source": [
    "# Create multiXrank object\n",
    "multilayer = generic_multicell.Multixrank(\n",
    "    restart_proba=restart_proba\n",
    ")\n",
    "\n",
    "# Run random walk with restart\n",
    "results = multilayer.random_walk_rank()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "575705cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>multiplex</th>\n",
       "      <th>node</th>\n",
       "      <th>layer</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cell_communication</td>\n",
       "      <td>Abca1-B_cell</td>\n",
       "      <td>cell_communication</td>\n",
       "      <td>1.100054e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cell_communication</td>\n",
       "      <td>Abca1-Macrophage</td>\n",
       "      <td>cell_communication</td>\n",
       "      <td>1.607757e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cell_communication</td>\n",
       "      <td>Abca1-NK_cell</td>\n",
       "      <td>cell_communication</td>\n",
       "      <td>9.655072e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cell_communication</td>\n",
       "      <td>Abca1-T_cell_CD4</td>\n",
       "      <td>cell_communication</td>\n",
       "      <td>7.274640e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cell_communication</td>\n",
       "      <td>Abca1-T_cell_CD8</td>\n",
       "      <td>cell_communication</td>\n",
       "      <td>9.304301e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10163</th>\n",
       "      <td>T_cell_CD8_grn</td>\n",
       "      <td>Zxdc::T_cell_CD8</td>\n",
       "      <td>gene</td>\n",
       "      <td>4.349167e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10164</th>\n",
       "      <td>T_cell_CD8_grn</td>\n",
       "      <td>Zyg11b::T_cell_CD8</td>\n",
       "      <td>gene</td>\n",
       "      <td>3.027501e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10165</th>\n",
       "      <td>T_cell_CD8_grn</td>\n",
       "      <td>Zyx::T_cell_CD8</td>\n",
       "      <td>gene</td>\n",
       "      <td>7.329956e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10166</th>\n",
       "      <td>T_cell_CD8_grn</td>\n",
       "      <td>Zzef1::T_cell_CD8</td>\n",
       "      <td>gene</td>\n",
       "      <td>2.635670e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10167</th>\n",
       "      <td>T_cell_CD8_grn</td>\n",
       "      <td>Zzz3::T_cell_CD8</td>\n",
       "      <td>gene</td>\n",
       "      <td>1.374288e-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67631 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                multiplex                node               layer  \\\n",
       "0      cell_communication        Abca1-B_cell  cell_communication   \n",
       "1      cell_communication    Abca1-Macrophage  cell_communication   \n",
       "2      cell_communication       Abca1-NK_cell  cell_communication   \n",
       "3      cell_communication    Abca1-T_cell_CD4  cell_communication   \n",
       "4      cell_communication    Abca1-T_cell_CD8  cell_communication   \n",
       "...                   ...                 ...                 ...   \n",
       "10163      T_cell_CD8_grn    Zxdc::T_cell_CD8                gene   \n",
       "10164      T_cell_CD8_grn  Zyg11b::T_cell_CD8                gene   \n",
       "10165      T_cell_CD8_grn     Zyx::T_cell_CD8                gene   \n",
       "10166      T_cell_CD8_grn   Zzef1::T_cell_CD8                gene   \n",
       "10167      T_cell_CD8_grn    Zzz3::T_cell_CD8                gene   \n",
       "\n",
       "              score  \n",
       "0      1.100054e-06  \n",
       "1      1.607757e-05  \n",
       "2      9.655072e-07  \n",
       "3      7.274640e-07  \n",
       "4      9.304301e-07  \n",
       "...             ...  \n",
       "10163  4.349167e-11  \n",
       "10164  3.027501e-11  \n",
       "10165  7.329956e-11  \n",
       "10166  2.635670e-10  \n",
       "10167  1.374288e-10  \n",
       "\n",
       "[67631 rows x 4 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "60e2a2d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages/recon/explore/recon.py:570: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/pasteur/appa/homes/rtrimbou/miniconda3/envs/snakemake/envs/recon-dev/lib/python3.10/site-packages/recon/explore/recon.py:570: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>celltype</th>\n",
       "      <th>B_cell</th>\n",
       "      <th>Macrophage</th>\n",
       "      <th>NK_cell</th>\n",
       "      <th>T_cell_CD4</th>\n",
       "      <th>T_cell_CD8</th>\n",
       "      <th>pDC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A3galt2</th>\n",
       "      <td>2.018279e-12</td>\n",
       "      <td>2.740803e-08</td>\n",
       "      <td>1.228604e-11</td>\n",
       "      <td>5.891528e-12</td>\n",
       "      <td>6.195689e-13</td>\n",
       "      <td>8.664049e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4galt</th>\n",
       "      <td>1.387521e-11</td>\n",
       "      <td>1.880820e-07</td>\n",
       "      <td>1.472483e-11</td>\n",
       "      <td>6.816724e-12</td>\n",
       "      <td>6.870771e-12</td>\n",
       "      <td>1.111332e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aa467197</th>\n",
       "      <td>9.843796e-12</td>\n",
       "      <td>1.202237e-06</td>\n",
       "      <td>4.097082e-11</td>\n",
       "      <td>2.940558e-11</td>\n",
       "      <td>2.343176e-11</td>\n",
       "      <td>1.974579e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aaas</th>\n",
       "      <td>2.592065e-11</td>\n",
       "      <td>1.515622e-06</td>\n",
       "      <td>5.045680e-11</td>\n",
       "      <td>4.397842e-11</td>\n",
       "      <td>3.124886e-11</td>\n",
       "      <td>6.148052e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aacs</th>\n",
       "      <td>9.098847e-11</td>\n",
       "      <td>9.323522e-06</td>\n",
       "      <td>2.129388e-10</td>\n",
       "      <td>1.335567e-10</td>\n",
       "      <td>1.056153e-10</td>\n",
       "      <td>1.560250e-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "celltype        B_cell    Macrophage       NK_cell    T_cell_CD4  \\\n",
       "gene                                                               \n",
       "A3galt2   2.018279e-12  2.740803e-08  1.228604e-11  5.891528e-12   \n",
       "A4galt    1.387521e-11  1.880820e-07  1.472483e-11  6.816724e-12   \n",
       "Aa467197  9.843796e-12  1.202237e-06  4.097082e-11  2.940558e-11   \n",
       "Aaas      2.592065e-11  1.515622e-06  5.045680e-11  4.397842e-11   \n",
       "Aacs      9.098847e-11  9.323522e-06  2.129388e-10  1.335567e-10   \n",
       "\n",
       "celltype    T_cell_CD8           pDC  \n",
       "gene                                  \n",
       "A3galt2   6.195689e-13  8.664049e-13  \n",
       "A4galt    6.870771e-12  1.111332e-11  \n",
       "Aa467197  2.343176e-11  1.974579e-11  \n",
       "Aaas      3.124886e-11  6.148052e-11  \n",
       "Aacs      1.056153e-10  1.560250e-10  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Format results as gene profiles per cell type\n",
    "cell_type_profiles = recon.explore.format_multicell_results(\n",
    "    multicell_multixrank_results=results,\n",
    "    celltypes=celltypes,\n",
    "    keep_layers=\"gene\"\n",
    ")\n",
    "\n",
    "cell_type_profiles.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "302b48b7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d096063",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "recon-dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
