{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1e98cdcd",
   "metadata": {},
   "source": [
    "# grn - Inferring GRN through ReCoN - HuMMuS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dc95caa",
   "metadata": {},
   "source": [
    "⚠️ Checking that bedtools is available"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "30efc2b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bedtools is a powerful toolset for genome arithmetic.\n",
      "\n",
      "Version:   v2.31.1\n",
      "About:     developed in the quinlanlab.org and by many contributors worldwide.\n",
      "Docs:      http://bedtools.readthedocs.io/\n",
      "Code:      https://github.com/arq5x/bedtools2\n",
      "Mail:      https://groups.google.com/forum/#!forum/bedtools-discuss\n",
      "\n",
      "Usage:     bedtools <subcommand> [options]\n",
      "\n",
      "The bedtools sub-commands include:\n",
      "\n",
      "[ Genome arithmetic ]\n",
      "    intersect     Find overlapping intervals in various ways.\n",
      "    window        Find overlapping intervals within a window around an interval.\n",
      "    closest       Find the closest, potentially non-overlapping interval.\n",
      "    coverage      Compute the coverage over defined intervals.\n",
      "    map           Apply a function to a column for each overlapping interval.\n",
      "    genomecov     Compute the coverage over an entire genome.\n",
      "    merge         Combine overlapping/nearby intervals into a single interval.\n",
      "    cluster       Cluster (but don't merge) overlapping/nearby intervals.\n",
      "    complement    Extract intervals _not_ represented by an interval file.\n",
      "    shift         Adjust the position of intervals.\n",
      "    subtract      Remove intervals based on overlaps b/w two files.\n",
      "    slop          Adjust the size of intervals.\n",
      "    flank         Create new intervals from the flanks of existing intervals.\n",
      "    sort          Order the intervals in a file.\n",
      "    random        Generate random intervals in a genome.\n",
      "    shuffle       Randomly redistribute intervals in a genome.\n",
      "    sample        Sample random records from file using reservoir sampling.\n",
      "    spacing       Report the gap lengths between intervals in a file.\n",
      "    annotate      Annotate coverage of features from multiple files.\n",
      "\n",
      "[ Multi-way file comparisons ]\n",
      "    multiinter    Identifies common intervals among multiple interval files.\n",
      "    unionbedg     Combines coverage intervals from multiple BEDGRAPH files.\n",
      "\n",
      "[ Paired-end manipulation ]\n",
      "    pairtobed     Find pairs that overlap intervals in various ways.\n",
      "    pairtopair    Find pairs that overlap other pairs in various ways.\n",
      "\n",
      "[ Format conversion ]\n",
      "    bamtobed      Convert BAM alignments to BED (& other) formats.\n",
      "    bedtobam      Convert intervals to BAM records.\n",
      "    bamtofastq    Convert BAM records to FASTQ records.\n",
      "    bedpetobam    Convert BEDPE intervals to BAM records.\n",
      "    bed12tobed6   Breaks BED12 intervals into discrete BED6 intervals.\n",
      "\n",
      "[ Fasta manipulation ]\n",
      "    getfasta      Use intervals to extract sequences from a FASTA file.\n",
      "    maskfasta     Use intervals to mask sequences from a FASTA file.\n",
      "    nuc           Profile the nucleotide content of intervals in a FASTA file.\n",
      "\n",
      "[ BAM focused tools ]\n",
      "    multicov      Counts coverage from multiple BAMs at specific intervals.\n",
      "    tag           Tag BAM alignments based on overlaps with interval files.\n",
      "\n",
      "[ Statistical relationships ]\n",
      "    jaccard       Calculate the Jaccard statistic b/w two sets of intervals.\n",
      "    reldist       Calculate the distribution of relative distances b/w two files.\n",
      "    fisher        Calculate Fisher statistic b/w two feature files.\n",
      "\n",
      "[ Miscellaneous tools ]\n",
      "    overlap       Computes the amount of overlap from two intervals.\n",
      "    igv           Create an IGV snapshot batch script.\n",
      "    links         Create a HTML page of links to UCSC locations.\n",
      "    makewindows   Make interval \"windows\" across a genome.\n",
      "    groupby       Group by common cols. & summarize oth. cols. (~ SQL \"groupBy\")\n",
      "    expand        Replicate lines based on lists of values in columns.\n",
      "    split         Split a file into multiple files with equal records or base pairs.\n",
      "    summary       Statistical summary of intervals in a file.\n",
      "\n",
      "[ General Parameters ]\n",
      "     --cram-ref    Reference used by a CRAM input\n",
      "\n",
      "[ General help ]\n",
      "    --help        Print this help menu.\n",
      "    --version     What version of bedtools are you using?.\n",
      "    --contact     Feature requests, bugs, mailing lists, etc.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!bedtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8c22f5ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import hummuspy.loader\n",
    "import circe as ci\n",
    "import recon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "47ea9f71",
   "metadata": {},
   "outputs": [],
   "source": [
    "import muon as mu\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dddd89ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import data\n",
    "mudata = mu.read_h5mu(\"circe_benchmark/data/datasets/pbmc10x/pbmc10x.h5mu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "53529504",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna = mudata.mod['rna'][:, :1000]\n",
    "atac = mudata.mod['atac'][:, :5000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d096d99a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tfs_list = hummuspy.loader.load_tfs(\"human_tfs_r_hummus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e8f63427",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating TF-to-gene importance\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running using 15 cores:   0%|          | 0/1000 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running using 15 cores: 100%|██████████| 1000/1000 [00:05<00:00, 195.65it/s]\n"
     ]
    }
   ],
   "source": [
    "rna_network = recon.infer_grn.compute_rna_network(rna, tf_names=tfs_list, n_cpu=15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11fab776",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "38daafb731d44b86bcc6406083b310ee",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8a39cc8ae608472797038d0872bca14c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating co-accessibility scores...\n",
      "Concatenating results from all chromosomes...\n"
     ]
    }
   ],
   "source": [
    "atac.var_names = atac.var_names.str.replace(\"-\", \"_\")\n",
    "atac.var_names = atac.var_names.str.replace(\":\", \"_\")\n",
    "\n",
    "atac = ci.add_region_infos(atac)\n",
    "\n",
    "ci.compute_atac_network(atac, njobs=20)\n",
    "\n",
    "atac_network = ci.extract_atac_links(atac, key='atac_network', )\n",
    "atac_network = atac_network.rename(columns={\n",
    "    \"Peak1\": \"source\",\n",
    "    \"Peak2\": \"target\",\n",
    "    \"score\": \"weight\"\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4358ea03",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_network = recon.infer_grn.compute_tf_network(rna, tfs_list=tfs_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "db25db56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TFAP2E_TF</td>\n",
       "      <td>fake_TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HES2_TF</td>\n",
       "      <td>fake_TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ID3_TF</td>\n",
       "      <td>fake_TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TAL1_TF</td>\n",
       "      <td>fake_TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HES4_TF</td>\n",
       "      <td>fake_TF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      source   target\n",
       "0  TFAP2E_TF  fake_TF\n",
       "1    HES2_TF  fake_TF\n",
       "2     ID3_TF  fake_TF\n",
       "3    TAL1_TF  fake_TF\n",
       "4    HES4_TF  fake_TF"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_network.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c9c9216",
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_rna_links = recon.infer_grn.compute_atac_to_rna_links(atac, rna, ref_genome=\"hg19\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "145fda2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Peaks before filtering:  5000\n",
      "Peaks with invalid chr_name:  0\n",
      "Peaks with invalid length:  0\n",
      "Peaks after filtering:  5000\n",
      "No motif data entered. Loading default motifs for your species ...\n",
      " Default motif for vertebrate: gimme.vertebrate.v5.0. \n",
      " For more information, please see https://gimmemotifs.readthedocs.io/en/master/overview.html \n",
      "\n",
      "Initiating scanner... \n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-09-24 10:46:00,616 - DEBUG - using background: genome hg19 with size 200\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating FPR-based threshold. This step may take substantial time when you load a new ref-genome. It will be done quicker on the second time. \n",
      "\n",
      "Motif scan started .. It may take long time.\n",
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5c4d20e8902244a6b675d0cf810c2f13",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Scanning:   0%|          | 0/5000 [00:00<?, ? sequences/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering finished: 702652 -> 156632\n",
      "1. Converting scanned results into one-hot encoded dataframe.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ffad85e4997248de8384f97c7f0cf65c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4962 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2. Converting results into dictionaries.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "90603f54112a4de78d13374ebecd523d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e5c0c1356f174d84babd59c6e9359a49",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1093 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tf_atac_links = recon.infer_grn.compute_tf_to_atac_links(atac, ref_genome=\"hg19\", n_cpus=30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b2b63e5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_atac_links = tf_atac_links[tf_atac_links['source'].str.replace(\"_TF\", \"\").isin(tfs_list)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "bd2ab3eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NEUROD1_TF</td>\n",
       "      <td>chr1_1000608_1001108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BHLHA15_TF</td>\n",
       "      <td>chr1_1000608_1001108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ATOH1_TF</td>\n",
       "      <td>chr1_1000608_1001108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>OLIG2_TF</td>\n",
       "      <td>chr1_1000608_1001108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PTF1A_TF</td>\n",
       "      <td>chr1_1000608_1001108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2999838</th>\n",
       "      <td>SP7_TF</td>\n",
       "      <td>chr1_999910_1000410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2999842</th>\n",
       "      <td>SP5_TF</td>\n",
       "      <td>chr1_999910_1000410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2999843</th>\n",
       "      <td>SP8_TF</td>\n",
       "      <td>chr1_999910_1000410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2999848</th>\n",
       "      <td>SP6_TF</td>\n",
       "      <td>chr1_999910_1000410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2999855</th>\n",
       "      <td>MYC_TF</td>\n",
       "      <td>chr1_999910_1000410</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1366039 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             source                target\n",
       "0        NEUROD1_TF  chr1_1000608_1001108\n",
       "2        BHLHA15_TF  chr1_1000608_1001108\n",
       "8          ATOH1_TF  chr1_1000608_1001108\n",
       "9          OLIG2_TF  chr1_1000608_1001108\n",
       "11         PTF1A_TF  chr1_1000608_1001108\n",
       "...             ...                   ...\n",
       "2999838      SP7_TF   chr1_999910_1000410\n",
       "2999842      SP5_TF   chr1_999910_1000410\n",
       "2999843      SP8_TF   chr1_999910_1000410\n",
       "2999848      SP6_TF   chr1_999910_1000410\n",
       "2999855      MYC_TF   chr1_999910_1000410\n",
       "\n",
       "[1366039 rows x 2 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_atac_links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "dcc77be3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TF_0\n",
      "ATAC_0\n",
      "RNA_0\n",
      "Initializing Dask cluster with 20 workers...\n",
      "http://192.168.152.117:8787/status\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running RWR per seed: 100%|██████████| 29/29 [00:02<00:00, 13.07it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building Dask DataFrame graph with delayed parallel tasks...\n"
     ]
    }
   ],
   "source": [
    "grn = recon.infer_grn.generate_grn(\n",
    "    rna_network=rna_network,\n",
    "    atac_network=atac_network,\n",
    "    tf_network=tf_network,\n",
    "    atac_to_rna_links=atac_rna_links,\n",
    "    tf_to_atac_links=tf_atac_links,\n",
    "    n_jobs=20\n",
    ")\n",
    "grn = grn[grn.layer == 'RNA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "ae23cde1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>layer</th>\n",
       "      <th>target</th>\n",
       "      <th>path_layer</th>\n",
       "      <th>score</th>\n",
       "      <th>seed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3639</th>\n",
       "      <td>883</td>\n",
       "      <td>RNA</td>\n",
       "      <td>TMEM35B</td>\n",
       "      <td>RNA_0</td>\n",
       "      <td>1.130846e-04</td>\n",
       "      <td>PAX7_TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3645</th>\n",
       "      <td>440</td>\n",
       "      <td>RNA</td>\n",
       "      <td>LINC01355</td>\n",
       "      <td>RNA_0</td>\n",
       "      <td>1.130419e-04</td>\n",
       "      <td>PAX7_TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>658</td>\n",
       "      <td>RNA</td>\n",
       "      <td>PLEKHG5</td>\n",
       "      <td>RNA_0</td>\n",
       "      <td>1.030718e-04</td>\n",
       "      <td>JUN_TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3657</th>\n",
       "      <td>139</td>\n",
       "      <td>RNA</td>\n",
       "      <td>CELA2B</td>\n",
       "      <td>RNA_0</td>\n",
       "      <td>7.939168e-05</td>\n",
       "      <td>ZNF684_TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3664</th>\n",
       "      <td>6</td>\n",
       "      <td>RNA</td>\n",
       "      <td>ACOT7</td>\n",
       "      <td>RNA_0</td>\n",
       "      <td>7.936830e-05</td>\n",
       "      <td>ZNF684_TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3317</th>\n",
       "      <td>856</td>\n",
       "      <td>RNA</td>\n",
       "      <td>TAS1R1</td>\n",
       "      <td>RNA_0</td>\n",
       "      <td>6.172440e-08</td>\n",
       "      <td>PAX7_TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3431</th>\n",
       "      <td>596</td>\n",
       "      <td>RNA</td>\n",
       "      <td>NOL9</td>\n",
       "      <td>RNA_0</td>\n",
       "      <td>5.801554e-08</td>\n",
       "      <td>FOXO6_TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3432</th>\n",
       "      <td>856</td>\n",
       "      <td>RNA</td>\n",
       "      <td>TAS1R1</td>\n",
       "      <td>RNA_0</td>\n",
       "      <td>5.801554e-08</td>\n",
       "      <td>FOXO6_TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3439</th>\n",
       "      <td>596</td>\n",
       "      <td>RNA</td>\n",
       "      <td>NOL9</td>\n",
       "      <td>RNA_0</td>\n",
       "      <td>5.736415e-08</td>\n",
       "      <td>ZNF684_TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3440</th>\n",
       "      <td>856</td>\n",
       "      <td>RNA</td>\n",
       "      <td>TAS1R1</td>\n",
       "      <td>RNA_0</td>\n",
       "      <td>5.736415e-08</td>\n",
       "      <td>ZNF684_TF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2929 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index layer     target path_layer         score       seed\n",
       "3639    883   RNA    TMEM35B      RNA_0  1.130846e-04    PAX7_TF\n",
       "3645    440   RNA  LINC01355      RNA_0  1.130419e-04    PAX7_TF\n",
       "996     658   RNA    PLEKHG5      RNA_0  1.030718e-04     JUN_TF\n",
       "3657    139   RNA     CELA2B      RNA_0  7.939168e-05  ZNF684_TF\n",
       "3664      6   RNA      ACOT7      RNA_0  7.936830e-05  ZNF684_TF\n",
       "...     ...   ...        ...        ...           ...        ...\n",
       "3317    856   RNA     TAS1R1      RNA_0  6.172440e-08    PAX7_TF\n",
       "3431    596   RNA       NOL9      RNA_0  5.801554e-08   FOXO6_TF\n",
       "3432    856   RNA     TAS1R1      RNA_0  5.801554e-08   FOXO6_TF\n",
       "3439    596   RNA       NOL9      RNA_0  5.736415e-08  ZNF684_TF\n",
       "3440    856   RNA     TAS1R1      RNA_0  5.736415e-08  ZNF684_TF\n",
       "\n",
       "[2929 rows x 6 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "8bfd6948",
   "metadata": {},
   "outputs": [],
   "source": [
    "grn.to_csv(\"recon_hummus_grn.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43f19203",
   "metadata": {},
   "source": [
    "This work uses **ReCoN**’s wrapper of different packages, which are all cited in the Methods section:\n",
    "\n",
    "- The **RNA layer** uses [**arboreto¹**](#arboreto)  \n",
    "- The **ATAC layer** uses [**CIRCE²**](#circe)  \n",
    "- The **TF-ATAC** and **ATAC-RNA bipartites** use [**CellOracle³**](#celloracle)  \n",
    "- The exploration of the **multilayer** uses [**HuMMuS⁴**](#hummus) and [**MultiXrank⁵**](#multixrank)  \n",
    "\n",
    "---\n",
    "\n",
    "### References\n",
    "<a id=\"arboreto\"></a>\n",
    "[1] Moerman, T., Aibar, S., Bravo González-Blas, C., Simm, J., Moreau, Y., Aerts, J., & Aerts, S. (2019).  \n",
    "**GRNBoost2 and Arboreto: efficient and scalable inference of gene regulatory networks.**  \n",
    "*Bioinformatics*, 35(12), 2159–2161. doi:[10.1093/bioinformatics/bty916](https://doi.org/10.1093/bioinformatics/bty916)\n",
    "\n",
    "<a id=\"circe\"></a>\n",
    "[2] Trimbour, R., Saez Rodriguez J., Cantini L. (2025).\n",
    "**Circe: Co-accessibility network from ATAC-seq data in Python (software).**  \n",
    "Version 0.3.6. [GitHub](https://github.com/cantinilab/Circe)\n",
    "\n",
    "<a id=\"celloracle\"></a>\n",
    "[3] Kamimoto, K., Stringa, B., Hoffmann, C. M., Jindal, K., Solnica-Krezel, L., Morris, S. A., et al. (2023).  \n",
    "**Dissecting cell identity via network inference and in silico gene perturbation.**  \n",
    "*Nature*, 614, 742–751. doi:[10.1038/s41586-022-05688-9](https://doi.org/10.1038/s41586-022-05688-9)\n",
    "\n",
    "<a id=\"hummus\"></a>\n",
    "[4] Trimbour, R., Deutschmann I. M., Cantini L. (2024).  \n",
    "**Molecular mechanisms reconstruction from single-cell multi-omics data with HuMMuS.**  \n",
    "*Bioinformatics*, 40(5), btae143. doi:[10.1093/bioinformatics/btae143](https://doi.org/10.1093/bioinformatics/btae143)\n",
    "\n",
    "<a id=\"multixrank\"></a>\n",
    "[5] Baptista, A., González, A., & Baudot, A. (2022).  \n",
    "**Universal multilayer network exploration by random walk with restart.**  \n",
    "*Communications Physics*, 5, 170. doi:[10.1038/s42005-022-00937-9](https://doi.org/10.1038/s42005-022-00937-9)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1b033ed",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "celloracle_env",
   "language": "python",
   "name": "celloracle_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
